### PR TITLE
Bridge HeaplessBigInt onto the num_traits foundation traits

### PR DESCRIPTION
**Pull request by kaidokert**
First PR of the HeaplessBigInt feature-completion series (roadmap: notes/HEAPLESS_COMPLETION.md; gap inventory: notes/HEAPLESS_PARITY_INVENTORY.md — both gitignored).

**Why:** HeaplessBigInt satisfied *zero* `num_traits` bounds, so no generic `num_traits` consumer could accept it even though it already had the const-num-traits equivalents. This lands the foundation layer that `Num`/`PrimInt`/`Integer` all build on.

**What:**
- `const_num_traits::Bounded` — `min = 0`; `max` = the capacity bound (all `CAP` limbs saturated, `len = CAP`). This is the one answer `CAP` legitimately sets: the widest value the storage can hold, not a value width.
- `num_traits` `Zero`/`One`/`Bounded`/`ToPrimitive`/`FromPrimitive`/`NumCast` (`feature = "num-traits"`, personality-generic), thin forwards onto the const-num-traits impls plus a byte-based primitive extraction mirroring FixedUInt's bridge.

**Two carrier-specific wrinkles vs the FixedUInt bridge:**
- `num_traits::One` requires `Mul`, and heapless `Mul` needs `CarryingMul` on the word (FixedUInt widens via `DoubleWord`), so `One` carries that bound.
- `FromPrimitive::from_u64` trims trailing zero bytes before `from_le_bytes`, so a small value in a small-`CAP` carrier returns `None` on overflow rather than tripping the oversize assert — the trait contract is `None`, never panic.

**Tests:** identity/bounds, to_u64 round-trip + overflow, NumCast across widths, natural-width construction, and a value-for-value parity check against the same-width `FixedUInt` (the type this mirrors). Verified across default / --no-default-features / nightly; clippy clean.

Follow-ups in the series: PrimBits/PrimInt (carries the value-width decision — Not/count_zeros/etc. over `len·word_bits`) and num_integer::Integer (gcd/lcm/...).

## Summary by Sourcery

Bridge HeaplessBigInt onto the standard num_traits foundation traits so it can participate in generic numeric code.

New Features:
- Implement const_num_traits::Bounded for HeaplessBigInt to expose its capacity-based numeric bounds.
- Add a num-traits feature-gated bridge module wiring HeaplessBigInt into num_traits::Zero, One, Bounded, ToPrimitive, FromPrimitive, and NumCast.

Tests:
- Add tests covering Zero/One/Bounded behavior, u64 round-trip and overflow handling, NumCast across carrier widths, parity with FixedUInt behavior, and natural-width construction semantics.

## Summary by CodeRabbit

* **New Features**
  * Added optional `num-traits` compatibility for heapless big integers.
  * Added zero, one, minimum, and maximum value support.
  * Added conversion to and from unsigned 64-bit values, including overflow checks.
  * Added numeric casting support across compatible integer widths.

* **Tests**
  * Added coverage for numeric identities, bounds, conversions, casting, overflow handling, and precision behavior.

### CODE REVIEW

File: src/heapless/identities.rs [hunk 1]
```diff
@@ -146,3 +146,24 @@ impl<T: MachineWord, const CAP: usize, P: Personality> ConstZero for HeaplessBig
...
+    }
+
+    #[inline]
+    fn max_value() -> Self {
+        Self {
+            limbs: [<T as Bounded>::max_value(); CAP],
+            len: CAP as u16,
+            _p: PhantomData,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/identities.rs:171]: @gemini-code-assist Fixed in fe3a253 — max_value now evaluates `<Self as AssertCapFits>::CHECK` first, turning CAP > u16::MAX into a compile error instead of a silent len=0 truncation, consistent with the other shape-setting constructors.

File: src/heapless/num_traits_bridge.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+//!
+//! Thin forwards onto the const-num-traits identity/bounds impls plus
+//! byte-based primitive extraction, mirroring `FixedUInt`'s bridge so a
+//! `HeaplessBigInt` satisfies the classic `num_traits` bounds
+//! (`Zero`/`One`/`Bounded`/`NumCast`/`ToPrimitive`/`FromPrimitive`). All
+//! are personality-generic — no `Nct` gate, matching `FixedUInt`.
+
+use super::HeaplessBigInt;
+use crate::MachineWord;
+use const_num_traits::{Bounded, CarryingMul, ConstOne, ConstZero, Personality, Zero};
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/num_traits_bridge.rs:11]: @gemini-code-assist This one doesn't apply — the file has no `use num_traits::ToPrimitive` and it compiles/tests green as-is. `to_u64` is callable because the method comes into scope from the `X: num_traits::ToPrimitive` bound on the generic parameter, not from a `use`. Adding the import would be flagged unused, so leaving it out.

File: src/heapless/identities.rs [hunk 3]
```diff
@@ -146,3 +146,24 @@ impl<T: MachineWord, const CAP: usize, P: Personality> ConstZero for HeaplessBig
...
+    #[inline]
+    fn min_value() -> Self {
+        <Self as ConstZero>::ZERO
+    }
+
+    #[inline]
+    fn max_value() -> Self {
+        Self {
+            limbs: [<T as Bounded>::max_value(); CAP],
+            len: CAP as u16,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/identities.rs:168]: Fixed in fe3a253 — max_value now routes through AssertCapFits::CHECK like the other constructors, so CAP > u16::MAX is a compile error.

File: src/heapless/num_traits_bridge.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        <Self as Bounded>::max_value()
+    }
+}
+
+impl<T: MachineWord, const CAP: usize, P: Personality> num_traits::ToPrimitive
+    for HeaplessBigInt<T, CAP, P>
+{
+    fn to_i64(&self) -> Option<i64> {
+        // Unsigned carrier: mirror `FixedUInt`, which never claims `i64`.
+        None
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/num_traits_bridge.rs:53]: Intentional for now: this bridge deliberately mirrors FixedUInt's, which also returns None from to_i64. HeaplessBigInt is contracted to behave bit-for-bit like the same-width FixedUInt, so diverging to support signed casts here would break that parity. Broadening signed-primitive support is worth doing, but as a separate change applied to both carriers together.

File: src/heapless/num_traits_bridge.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        Some(ret)
+    }
+}
+
+impl<T: MachineWord, const CAP: usize, P: Personality> num_traits::FromPrimitive
+    for HeaplessBigInt<T, CAP, P>
+{
+    fn from_i64(_: i64) -> Option<Self> {
+        None
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/num_traits_bridge.rs:87]: Same as the to_i64 note: from_i64 returning None mirrors FixedUInt exactly, keeping the two carriers consistent. Non-negative signed-input support is a reasonable follow-up, but belongs on both types at once rather than diverging heapless here.

File: src/heapless/num_traits_bridge.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/num_traits_bridge.rs:53]: @coderabbitai make a follow-up ticket for this

File: src/heapless/num_traits_bridge.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/heapless/num_traits_bridge.rs:87]: @coderabbitai tag this one in #149  for followup too
